beefy-gadget/src/worker.rs

```diff
@@ -280,26 +384,28 @@ where
 				signature,
 			};
 
-			self.gossip_engine
-				.lock()
-				.gossip_message(topic::<B>(), message.encode(), false);
-
-			debug!(target: "beefy", "🥩 Sent vote message: {:?}", message);
+			let encoded_message = message.encode();
 
 			if let Some(metrics) = self.metrics.as_ref() {
 				metrics.beefy_gadget_votes.inc();
 			}
 
+			debug!(target: "beefy", "🥩 Sent vote message: {:?}", message);
+
 			self.handle_vote(
 				(message.commitment.payload, *message.commitment.block_number),
 				(message.id, message.signature),
 			);
-		}
 
-		self.best_finalized_block = *notification.header.number();
+			self.gossip_engine
+				.lock()
+				.gossip_message(topic::<B>(), encoded_message, false);
+		}
 	}
 
-	fn handle_vote(&mut self, round: (MmrRootHash, NumberFor<B>), vote: (P::Public, S)) {
+	fn handle_vote(&mut self, round: (MmrRootHash, NumberFor<B>), vote: (P::Public, P::Signature)) {
+		self.gossip_validator.note_round(round.1);
```

Comment:
Two things  that pop into my mind:
1. `handle_vote` is being called in 2 places (when we send the vote and when we receive vote from gossip). IMHO the `note_round` should only happen in the first case. It's safe now, cause we use it in the gossip validator anyway, but still I think it would be cleaner.
2. Non-voting workers will now never gossip anything. We should probably change the logic of `handle_finality_notification`, so that in case `should_vote_on` is `true` with `note_round` unconditionally.

Reply:
> They should be as they call handle_vote whenever they get a gossip message, and this will note the round in the gossip validator.

So that what confuses me. The `GossipValidator` seems to reject messages that have not been `noted` yet, and the only place where we `note_round` is here, so it will only be noted if the node is validating.

Reply:
The entry point for the gossip validator is the `validate` method, which only tries to decode the message. Returning `GossipValidationResult::ProcessAndKeep` will make sure the beefy listener will receive the message. `message_allowed` and `message_expired` are called later on when deciding whether to send a particular message to a given peer, or whether to prune the message from the gossip layer. Those methods will only be called after the beefy worker has processed the message and its round should already have been noted by then.

Reply:
Ah, that makes sense, if `message_allowed` is only called for outgoing messages we are good!